pkg/cmd/roachtest/tests/mixed_version_cdc.go

```diff
@@ -0,0 +1,318 @@
+				if err != nil {
+					t.Fatal(err)
+					return nil
+				}
+
+				partitionStr := strconv.Itoa(int(m.Partition))
+				// errors in the calls to the validator below could lead to
+				// transient failures if the cluster is upgrading, so we retry
+				// for 1 minute.
+				if len(m.Key) > 0 {
+					if err := retry.ForDuration(1*time.Minute, func() error {
+						return cmvt.validator.NoteRow(partitionStr, string(m.Key), string(m.Value), updated)
+					}); err != nil {
+						t.Fatal(err)
+						return nil
+					}
+				} else {
+					if err := retry.ForDuration(1*time.Minute, func() error {
+						return cmvt.validator.NoteResolved(partitionStr, resolved)
+					}); err != nil {
+						t.Fatal(err)
+						return nil
+					}
+					t.L().Printf("%d of %d resolved timestamps validated, latest is %s behind realtime",
+						cmvt.validator.NumResolvedWithRows, requestedResolved, timeutil.Since(resolved.GoTime()))
+
+					if cmvt.validator.NumResolvedWithRows >= requestedResolved {
+						break
+					}
+				}
+			}
+
+			close(cmvt.verifierDone)
+			return nil
+		})
+	}
+}
+
+// validatorVerify checks if the validator has found any issues at the
+// time the function is called.
+func (cmvt *cdcMixedVersionTester) validatorVerify() versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		if failures := cmvt.validator.Failures(); len(failures) > 0 {
+			t.Fatalf("validator failures:\n%s", strings.Join(failures, "\n"))
+		}
+	}
+}
+
+// createChangeFeed issues a call to the given node to create a change
+// feed for the warehouse table.
+func (cmvt *cdcMixedVersionTester) createChangeFeed(node int) versionStep {
+	return func(ctx context.Context, t test.Test, u *versionUpgradeTest) {
+		t.Status("creating changefeed")
+		db := u.conn(ctx, t, node)
+		cdcClusterSettings(t, sqlutils.MakeSQLRunner(db))
+
+		opts := []string{"updated", "resolved"}
+		if _, err := db.Exec(
+			fmt.Sprintf("CREATE CHANGEFEED FOR %s INTO $1 WITH %s", target, strings.Join(opts, ", ")),
+			cmvt.kafka.sinkURL(ctx),
+		); err != nil {
+			t.Fatal(err)
+		}
+	}
+}
+
+func runCDCMixedVersions(
+	ctx context.Context, t test.Test, c cluster.Cluster, buildVersion version.Version,
+) {
+	predecessorVersion, err := PredecessorVersion(buildVersion)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	tester := newCDCMixedVersionTester(ctx, t, c)
+	tester.StartKafka(t, c)
+	defer tester.Cleanup()
+
+	// sleep is a step added at different points in the test when we
+	// want the workload to have some time to run against a specific
+	// version of the database
+	sleep := sleepStep(10 * time.Second)
+
+	// An empty string will lead to the cockroach binary specified by flag
+	// `cockroach` to be used.
+	const mainVersion = ""
+	newVersionUpgradeTest(c,
+		uploadAndStartFromCheckpointFixture(tester.crdbNodes, predecessorVersion),
+		tester.setupVerifier(1),
+		tester.installAndStartTPCCWorkload("15m"),
```

Comment:
nit: have this method take time.Duration?

Reply:
Done.